src/compiler/scala/tools/nsc/typechecker/Infer.scala

```diff
@@ -49,12 +49,21 @@ trait Infer extends Checkable {
          (removeRepeated || numFormals != numArgs)
       && isVarArgTypes(formals1)
     )
-    def lastType = formals1.last.dealiasWiden.typeArgs.head
-    def expanded(n: Int) = (1 to n).toList map (_ => lastType)
 
-    if (expandLast)
-      formals1.init ::: expanded(numArgs - numFormals + 1)
-    else
+    if (expandLast) {
+      // extract the T from T*
+      val lastType = formals1.last.dealiasWiden.typeArgs.head
+
+      val n = numArgs - numFormals + 1
+      // Optimized version of: formals1.init ::: List.fill(n)(lastType)
+      val result = mutable.ListBuffer[Type]()
+      var fs = formals1
+      while ((fs ne Nil) && (fs.tail ne Nil)) {
+        result.addOne(fs.head)
+        fs = fs.tail
+      }
```

Comment:
If you want to be very micro about your optimizations, after this loop, `fs.head.dealiasWiden.typeArgs.head` is your `lastType`. Also `fs` should never be `Nil`.